From 06b01dfa4e29b5a6838dd9050f3e9418c5dc77f5 Mon Sep 17 00:00:00 2001
From: Sam James <sam@gentoo.org>
Date: Wed, 1 Dec 2021 14:58:40 +0000
Subject: [PATCH 011/196] clone-internal.c: make clone3 syscall optional

We're disabling clone3 for now _CONDITIONALLY_ (not by default) to allow
compatibility with applications using older Electron.

Use -DGENTOO_USE_CLONE3 to enable clone3 for now. In future, we will
revert back to always using clone3.

This was impacting e.g. Discord and Skype. This patch stops glibc from using
clone3 internally (which is the only real use of it) and falls back to the old
behaviour.

Specifically, we want https://github.com/electron/electron/pull/31091
to work its way downstream to various Electron applications.

https://bugs.gentoo.org/819045
https://bugs.gentoo.org/827386

https://git.launchpad.net/~ubuntu-core-dev/ubuntu/+source/glibc/plain/debian/patches/ubuntu/disable-clone3.patch

This is the same as the patch that was considered but ultimately rejected
for 2.34 because Docker got sorted out in time:
https://patchwork.ozlabs.org/project/glibc/patch/87eebkf8ph.fsf@oldenburg.str.redhat.com/.

[HH: updated for 2.38]
Signed-off-by: Sam James <sam@gentoo.org>
---
 sysdeps/unix/sysv/linux/clone-internal.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/sysdeps/unix/sysv/linux/clone-internal.c b/sysdeps/unix/sysv/linux/clone-internal.c
index a71effcbd3..60efe1f453 100644
--- a/sysdeps/unix/sysv/linux/clone-internal.c	2023-07-31 19:54:16.000000000 +0200
+++ b/sysdeps/unix/sysv/linux/clone-internal.c	2023-08-02 21:33:10.792227624 +0200
@@ -103,7 +103,7 @@ int
 __clone_internal (struct clone_args *cl_args,
 		  int (*func) (void *arg), void *arg)
 {
-#ifdef HAVE_CLONE3_WRAPPER
+#if defined(HAVE_CLONE3_WRAPPER) && defined(GENTOO_USE_CLONE3)
   int saved_errno = errno;
   int ret = __clone3_internal (cl_args, func, arg);
   if (ret != -1 || errno != ENOSYS)
